doc/operator-notes.md

```diff
@@ -30,3 +30,34 @@ In the second two cases, where there is a preexisting filesystem, Ignition's beh
 If `wipeFilesystem` is set to true, Ignition will always wipe any preexisting filesystem and create the desired filesystem. Note this will result in any data on the old filesystem being lost.
 
 If `wipeFilesystem` is set to false, Ignition will then attempt to reuse the existing filesystem. If the filesystem is of the correct type, has a matching label, and has a matching UUID, then Ignition will reuse the filesystem. If the label or UUID is not set in the Ignition config, they don't need to match for Ignition to reuse the filesystem. Any preexisting data will be left on the device and will be available to the installation. If the preexisting filesystem is *not* of the correct type, then Ignition will fail, and the machine will fail to boot.
+
+## Partition Reuse Semantics
+Similar to the filesystemd use semantics, there are three possibilties for partitions when Ignition runs
+
+- The is no preexisting partition
+- There is a preexisting partition of the correct size, starting sector, label, type_guid and guid.
+- There is a preexisting partition of an incorrect size, starting sector, label, type_guid or guid.
+
+In the first case, Ignition will always create the desired partition. Using a partition number of 0 specifies to use the next available partition number and will ensure there is no preexisting partition.
+
+In the second two cases, the wipePartition flag controls Ignition's behavior. If it is set to true, Ignition will first delete the partition with the specified number and then attempt to create the new partition. The partition number must be specified and non-zero if wipePartition is set to true.
```

Comment:
> How can I delete a partition outright?

You can't, hmmm. We could add another field for if we should skip creating a partition, but that feels a little like the old filesystem `create` semantics, just inverted. Alternatively, we could add a `none` (or something similar to that) keyword for `type_guid`. We were already planning on adding aliases for other common GUIDs and the `type_guid` field seems like the most logical field to overload:
 * It specifies the type of the partition; not having a partition is similar to the "type" of a partition.
 * It's a string, so we can pick a good name for the keyword
 * It's already going to have special keyword aliases.

> Does sgdisk enforce uniqueness of partition GUIDs? If not, should we? The GPT spec says, optimistically

Just tested this, no. We don't currently enforce it. I'm in favor of Validate() function that generates a warning, but not failing because of it.

Reply:
Regarding deleting things: We also can't delete a filesystem outright. We should probably add support for that as well doing whatever we decide to do here.

Reply:
I'm not wild about overloading any of the fields, but particularly that one.  "To delete a partition, set `"wipePartition": "true"` and `"type_guid": "none"`." doesn't seem like great UX.  We'd also be special-casing a keyword-to-UUID mapper to have completely different semantics in one particular case.

The problem with `create` is that it was imperative.  We could have a `shouldExist` field (but with a better name), which is not.  Not sure that's great either.

Complication: if I set partition 10 `shouldExist=false` (or `type_guid=none` for that matter) and then ask for a new partition 0, what partition number do I get?  Right now we let sgdisk pick a partition, but it won't know about the hole at 10.

> I'm in favor of Validate() function that generates a warning, but not failing because of it.

It's probably safe to assume that warnings in Ignition (vs. CT) will never be noticed.

> We also can't delete a filesystem outright.

Deleting a filesystem is not a thing.  Partitions are records that can be created or destroyed; filesystems are merely the current contents of a fixed set of blocks.  We could add support for running `wipefs` on a filesystem, but that's sufficiently obscure that I think we can skip for now.

Reply:
I'm fine with a `shouldExist` type of field. We should just error if `shouldExist` is false and they specified partition information since they are mutually exclusive. `shouldExist` isn't a terrible name; I can't think of anything better off the top of my head.

I went through and outlined all the combinations of partitions existing or not and the options. P.E. is if a partition is already there, S.E. is `shouldExist`
```
P.E.    S.E.    wipe?   Action
---------------------------------------------------
0       0       0       Do nothing
0       0       1       Do nothing
0       1       0       Create new part
0       1       1       Create new part
1       0       0       Fail (can't wipe, theres a part there that shouldn't be)
1       0       1       Delete partition
1       1       0       Check if new part matches, fail if it doesn't
1       1       1       Delete part, create new part
```
Basically `shouldExist` and the rest of the partition spec control what the partition should look like (including not existing) and `wipePartition` controls what action Ignition is allowed to take. Seem reasonable? Would a matrix like this be useful in the docs?

As for the complication, I just tested it. If the deletion is before the create in the `sgdisk` arguments, it will reuse the partition number it just freed, since sgdisk arguments are executed in order (this is documented in the man page, not a quirk of implementation). I think if we document the behavior it will be ok.

This however exposes another complication: it will _not_ reuse the same space on the disk if `start` is 0, because `sgdisk` specifies 0 to be the start of the largest block, not the "best fit", so the partition entries will be out of order. I don't know if that will be an issue and it's not an easy fix.

> It's probably safe to assume that warnings in Ignition (vs. CT) will never be noticed. 

CT will pick them up once a newer Ignition makes it into CT: CT runs `Validate()` on its generated output and displays those errors/warnings as well.

> We could add support for running wipefs on a partition

This is what I was refering to.

> but that's sufficiently obscure that I think we can skip for now.

Possible use case: I want to leave the root partition as is (maybe I have some special configuration via systemd units that run on first boot to do something special with it that ignition doesn't support) and I'm putting the root filesystem somewhere else. I need to get rid of the `ROOT` label. I could overwrite it with a new filesystem, but it seems silly to create something overtop something else just to delete it.

Reply:
I agree with the matrix, and also with including it somewhere in the docs.  The naming still seems problematic: will it be easy to remember the difference between `shouldExist` and `wipePartition`?

We can document the partition number 0 behavior but it doesn't seem ideal.  However, as you point out, our hands are largely tied: we don't have our own semantics for `0` in any of the fields, but inherit those of `sgdisk`.  We should have had our own semantics from the beginning, but that would have required Ignition to understand the current state of the partition table.

As discussed offline, CT `Validate` only helps for UUID collisions visible in the config, and not for collisions with existing UUIDs on disk.  Probably still useful to add the check though.

As also discussed offline, I'm okay with adding a filesystem type that runs `wipefs`, provided that it fails if a label or FS UUID are specified, but I'd like @crawford's input on whether that sort of feature is within Ignition's remit.

Reply:
With this, how would I express "make this partition (overwriting if necessary) if what's on disk doesn't match, use what's there if it does"?

Using the above matrix if a partition exists and the config describes that a partition should exist, then I have two options:
- wipe is false: fail if the partition doesn't match
- wipe is true: always wipe the partition

As a real world example of why I would want this behavior: I'm PXE booting and thus always running Ignition. The disks have random old setups on them from earlier installations. The first boot provision things so the partitions match what I want, and on following boots reuse the existing partitions without changing the Ignition config.

Reply:
You can't with this setup. We could make wipePartition a ternary type consisting of `never`, `always` and `whenNeeded`

```
P.E.    S.E.   wipe?   Action
---------------------------------------------------
0       0      never   Do nothing
0       0      always  Do nothing
0       0      needed  Do nothing

0       1      never   Create new part
0       1      always  Create new part
0       1      needed  Create new part

1       0      never   Fail (can't wipe, theres a part there that shouldn't be)
1       0      always  Delete partition
1       0      needed  Delete partition

1       1      never   Check if new part matches, fail if it doesn't
1       1      always  Delete part, create new part
1       1      needed  Check if new part matches, delete and create new part if it doesn't
```
Footnote: `whenNeeded` is a terrible name

Reply:
Additionally, we should use the logical complement of `shouldExist` so it doesn't need to be specified every time, the downside being finding a good name for it. Alternatively, break out everything but `wipePartitionEntry` and `number` into a sub-struct that's nullable. i.e:

```
# example creating a partition
{
  "wipePartitionEntry": "always",
  "number": 9,
  "attributes": {
    "label": "foobar",
    "size": 123456
  }
}

# example deleting a partition
{
  "wipePartitionEntry": "always",
  "number": 9,
}
```

As for names, I think `wipePartitionEntry` is fine.

Ideas for a better name for `shouldNotExist`: `isNonexistant`, `missing`, `noPartition`, `deleted`, `delete`, `erase`, `absent`, `absentFromPartitionTable` 

Ideas for `attributes` if we go that route: `info`, `partition`, `spec`, `partitionSpec`

Reply:
I was assuming `shouldExist` would be implemented as a reference to bool, nil treated as true.  Inverting the sense of the key seems confusing.

I'm not seeing an easy way to handle Derek's case.  If the user wants to recreate the partition with the same offset and length, but perhaps a different type GUID, `wipePartitionEntry` is ~fine (though the instance GUID would either need to be hardcoded, thus colliding across machines, or would change on every boot).  But if they're recreating the entire disk, most likely the partition boundaries would change, which means that they probably want a conditional `wipeTable` that only fires when any of the described partitions are incorrect.

It feels as though we're starting to push the limits of a declarative spec.

Reply:
I'm okay with `shouldExist` being a reference to a bool, and only a little hesitent because usually nil equates something false-like, but as long as we document it (both in docs and in source) I'm okay with it.

We can't use `wipePartitionEntry` as you described if start or size is 0, since it can move due to how `sgdisk` handles those.

I don't think Derek's user is recreating the entire disk, just modifying a partition or two.

Reply:
> I don't think Derek's user is recreating the entire disk, just modifying a partition or two.

The same basic observation still applies.  If the new partition isn't aligned with the old one, then in fact the user wants something moderately complicated to happen.

Reply:
Using the terminology of [this comment](https://github.com/coreos/ignition/pull/436/files#r137859774), do we actually need `always` mode?  Assuming that it doesn't make sense with a zero start/size, the only thing `always` would accomplish is to regenerate the instance GUID on each boot.

Reply:
To be sure I understand: recreating a partition that matches with the same attributes has no effect (overwrites the partition entry with the same entry), so you can safely recreate matching partitions, except the instance guid will change, thus `always` and `whenNeeded` are identical except `whenNeeded` doesn't change the instance guid.

Your idea is to keep `wipePartitionEntry` a boolean and have `true` be handled like the description for `whenNeeded`, yes?

One thing that `always` could accomplish is resizing an existing partition. `size` (or `start` for that matter) are weird because for matching 0 means "dont care" but for creating it means biggest size or start of biggest block. Without `always`, you can't resize a partition to be the biggest a disk would support. But then this isn't useful to PXE users would want a stable instance guid.

The real problem is `start` and `size` can't match on 0 and have it mean what sgdisk means. It might be possible to use `sgdisk --pretend` to see if deleting and recreating a partition with `size` or `start` 0 would change anything, but that means kind of hacky and error prone, especially since `sgdisk` doesn't have machine readable output.

Reply:
Yes, that was my proposal, but I see your point.

Reply:
If we want to be able to support PXE users and size/start/number 0, Ignition needs to understand what sgdisk will do, especially for partition numbers. The good news is `start`/`size` 0 and `number` 0 are seperate issues.

**Number 0**

Lets say partitions 1-4 are used, and the user specifies 4 should not exist and to create a new partition with number 0. That new partition will be created with number 4, since its the lowest number sgdisk can find. Next time Ignition runs it will delete and recreate partition 4, changing the uuid, which is is not what the user expects.

Ignition will need to handle picking the partition number when 0 is specified. This is surpisingly tricky however, since it needs to understand both what Ignition knows about and what is on the disk and depends on the order in which the partitions are created.

Alternatively, since we're going through a spec bump, we could discontinue support for number 0 altogether - it's only useful as a shortcut and in the case the user doesn't know what's on their disk, but also doesn't want to overwrite it, which I'm not sure is a real usecase.

**Start and Size 0**

The idea behind supporting matching a partition is too allow persistant partitions over Ignition runs. The partitions created from one run of the Ignition config must match when the Ignition config is run a second time (sidenote: we should add tests for this). For non-zero start/size, this is easy - just use `libblkid` to find the start/end boundries and compare them to the spec. 

The actual values `start` and `size` use when specified as 0 are deterministic if the partitions are created in the same order. 

Ignition should also support checking `size` 0 and `start` 0 with the following semantics: 
 * start/size 0 matches if it is the same as it would be if we wiped everything we're allowed and then created all of the specified partitions.
This can be implemented as follows (assumes sgdisk output is parseable)
 * Record the state of the partition table currently
 * Run `sgdisk --pretend ...` as if we were creating everything for this disk (assume nothing matches) and record that state
 * Compare the pre and post partitioning states to determine if the `start` or `size` changed. If it did, that partition does not match

We need to pretend to do the whole disk at once to ensure partition creation order is handled correctly.

Ignition also needs to state that partitions will be considered in order for a given disk, or else the final output state will not be deterministic (although this is *not* true for the disks themselves).

**wipePartitionEntry=always**
This would only be useful if we have a use case where someone wants to change instance guid every boot. I don't know if that would ever be useful.